FakeTensor speedup: split prep_args_for_hash and verify_args_for_hash

aorenste opened the pull request:
Verifying that the various inputs into the dispatch cache key are valid is somewhat expensive. Instead of verifying them every time we compute a cache key we do the least amount of work to create the cache key and then only verify the first time we try to insert that cache key into the cache.

5.1% FakeTensor perf win on the microbenchmark (23.6% cumulative)

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #125312
* #124226
* __->__ #124225

aorenste committed:
split prep_args_for_hash and verify_args_for_hash

[ghstack-poisoned]

aorenste committed:
Update on "split prep_args_for_hash and verify_args_for_hash"




[ghstack-poisoned]

aorenste committed:
Update on "split prep_args_for_hash and verify_args_for_hash"




[ghstack-poisoned]

aorenste committed:
Update on "split prep_args_for_hash and verify_args_for_hash"




[ghstack-poisoned]

aorenste committed:
Update on "split prep_args_for_hash and verify_args_for_hash"




[ghstack-poisoned]

aorenste committed:
Update on "split prep_args_for_hash and verify_args_for_hash"




[ghstack-poisoned]

aorenste committed:
Update on "split prep_args_for_hash and verify_args_for_hash"




[ghstack-poisoned]

aorenste committed:
Update on "split prep_args_for_hash and verify_args_for_hash"




[ghstack-poisoned]

aorenste committed:
Update on "split prep_args_for_hash and verify_args_for_hash"




[ghstack-poisoned]

aorenste committed:
Update on "split prep_args_for_hash and verify_args_for_hash"




[ghstack-poisoned]

aorenste committed:
Update on "split prep_args_for_hash and verify_args_for_hash"




[ghstack-poisoned]

aorenste committed:
Update on "FakeTensor speedup: split prep_args_for_hash and verify_args_for_hash"




[ghstack-poisoned]

aorenste committed:
Update on "FakeTensor speedup: split prep_args_for_hash and verify_args_for_hash"




[ghstack-poisoned]

aorenste committed:
Update on "FakeTensor speedup: split prep_args_for_hash and verify_args_for_hash"




[ghstack-poisoned]

aorenste committed:
Update on "FakeTensor speedup: split prep_args_for_hash and verify_args_for_hash"




[ghstack-poisoned]

masnesral commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -1084,8 +1116,13 @@ def _cache_key(
-        return _DispatchCacheKey(key_values)
+        ]
+        # Translate any FakeTensor args to metadata.
+        if args:
```
Interesting. Did you find this actually helps? I thought I remembered playing around with avoiding calls to _prep_args_for_hash and I couldn't eek out any more perf.

masnesral commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -1126,43 +1163,81 @@ def _validate_cache_key(
+                output.append(arg)
 
-        return tuple(result)
+    def _verify_args_for_hash(self, args: Any):
```
If I'm reading this correctly, I think _prep_args_for_hash and _verify_args_for_hash need to stay in sync such that any time the former would append UNHASHABLE, the latter should raise _BypassDispatchCache. Is that right? I'm wondering if we could just note that we found an unhashable object during key calculation so that we don't have to re-check in `_validate_cache_key`?

aorenste replied:
It shouldn't be a perf issue - we'll only do the `_validate_cache_key` check once. After that the failure will be cached.

masnesral replied:
My question was more wondering if we could avoid having to keep these two sections of code in sync in case we add more conditions in the future.

aorenste committed:
Update on "FakeTensor speedup: split prep_args_for_hash and verify_args_for_hash"

Verifying that the various inputs into the dispatch cache key are valid is somewhat expensive. Instead of verifying them every time we compute a cache key we do the least amount of work to create the cache key and then only verify the first time we try to insert that cache key into the cache.

5.1% FakeTensor perf win on the microbenchmark (23.6% cumulative)




[ghstack-poisoned]

aorenste committed:
Update on "FakeTensor speedup: split prep_args_for_hash and verify_args_for_hash"

Verifying that the various inputs into the dispatch cache key are valid is somewhat expensive. Instead of verifying them every time we compute a cache key we do the least amount of work to create the cache key and then only verify the first time we try to insert that cache key into the cache.

5.1% FakeTensor perf win on the microbenchmark (23.6% cumulative)




[ghstack-poisoned]
